### ISSUE TITLE
Configure-build Macaulay2/e dependencies missing

### ISSUE THREAD

**Issue by MichaelABurr** [Macaulay2/M2#1990]
It appears that some dependencies for files in the Macaulay2/e directory may be missing.  

When I added or removed some virtual functions to Macaulay2/e/ring.hpp, I introduced errors or crashes (for example, removing the line for is_RRR() and running the command QQ[x] caused a hard crash).  It appears that the virtual function lookup tables were not updated after the change, which caused the wrong functions to be called.  

Cleaning the d and e directories corrected the error, but required manual intervention.

**Comment by DanGrayson** [comment]
Those *.dep dependency files are computed on the fly, so I wonder if there's a bug in gcc that makes the computation incorrect.

**Comment by mahrud** [comment]
> Cleaning the d and e directories corrected the error, but required manual intervention.

Do you know which one solved it? Is `is_RRR` used somewhere in the interpreter?

**Comment by MichaelABurr** [comment]
> > Cleaning the d and e directories corrected the error, but required manual intervention.
> 
> Do you know which one solved it? Is `is_RRR` used somewhere in the interpreter?

I don't know which solved the issue.  I did a grep and found symbols for is_RRR in binaries in both the d and e directories, so I cleaned them both.  is_RRR isn't used anywhere (as far as I can tell), but by removing it, I think that there was a (partial) shift in the virtual function lookup table, so the wrong functions were being called.

**Comment by MichaelABurr** [comment]
> Those *.dep dependency files are computed on the fly, so I wonder if there's a bug in gcc that makes the computation incorrect.

I'm using clang and not gcc, but the same question certainly applies.

**Comment by DanGrayson** [comment]
Actually, those computed *.dep files are just in the e directory.  The d directory, on the other hand, has dependencies set by humans in the makefiles.  So it's more likely that one of those is missing.

**Comment by DanGrayson** [comment]
PS: an experiment would show whether cleaning just the d directory works.

**Comment by MichaelABurr** [comment]
> PS: an experiment would show whether cleaning just the d directory works.

Cleaning only the d directory is not sufficient.

**Comment by mahrud** [comment]
> Cleaning only the d directory is not sufficient.

That's comforting. I think rather than adding Makefile or CMake dependencies, we should adjust the header files so the compiler knows when to recompile a source.

The test you run is simply deleting the `is_RRR` line in `ring.hpp` after compiling once, correct? Any other changes?

**Comment by mikestillman** [comment]
I think there is a chance that it is related to whether we include files via e.g.  "../f4/a.hpp" or "a.hpp" (in e.g. f4 directory).  We should do an experiment, but it seems one of these might work, and the other not.  But I didn't do the experiment, so I'm not completely sure.

**Comment by antonleykin** [comment]
`e/Makefile.files` exists... but doesn't get copied to the BUILD.
(... unlike `d/Makefile.files` that exists on both ends.)

**Comment by MichaelABurr** [comment]
> The test you run is simply deleting the `is_RRR` line in `ring.hpp` after compiling once, correct? Any other changes?

The steps to produce the error (on my system):
Step 1: Do a full configure-build
Step 2: Comment out is_RRR() in Macaulay2/e/ring.hpp
Step 3: Recompile (gmake in the build directory)
Step 4: Open M2 and run QQ[x] -- crash

**Comment by mahrud** [comment]
> I think there is a chance that it is related to whether we include files via e.g. "../f4/a.hpp" or "a.hpp" (in e.g. f4 directory). We should do an experiment, but it seems one of these might work, and the other not. But I didn't do the experiment, so I'm not completely sure.

This is not an issue anywhere in the engine. I've gone through and made sure of this.

**Comment by MichaelABurr** [comment]
It appears that the .o files in the subdirectories of the e directory are not being rebuilt.  This looks to be an issue with the Makefiles/dep files being off in the directory structure.

**Comment by mikestillman** [comment]
@DanGrayson I went through this with @MichaelABurr , and the .dep files are a problem.  Perhaps we can work this through on Tuesday when we meet?  It is an issue involving the subdirectories in the e directory, and .dep files and the Makefiles.

**Comment by DanGrayson** [comment]
Okay, but maybe at a different time -- I'm on Pacific time here.

**Comment by MichaelABurr** [comment]
> @DanGrayson I went through this with @MichaelABurr , and the .dep files are a problem. Perhaps we can work this through on Tuesday when we meet? It is an issue involving the subdirectories in the e directory, and .dep files and the Makefiles.

Adapting @mikestillman 's suggestion, I made the following changes to Macaulay2/e/Makefile.common.in in the source directory:

The four changes are in bold (two in each makefile command),
1) Add -MT"$@"
2) Change the .o to .dep in the sed command.

<pre>
%.dep: %.cpp
#	$(WHY)
	$(SHOW)"MKDEP $*.cpp"
	$(HIDE)$(MKDIR_P) `dirname "$@"`
	$(HIDE)@CXX@ <b>-MT"$@"</b> -MM -MP $(CPPFLAGS) $(CXXFLAGS) -DDEPENDS=1 $< >$*-tmp.dep && <$*-tmp.dep sed 's=\($*\)\.<b>dep</b>[ :]*=\1.o $@ : =' >$@ && rm $*-tmp.dep

%.dep: %.c
#	$(WHY)
	$(SHOW)"MKDEP $*.c"
	$(HIDE)@CC@ <b>-MT"$@"</b> -MM -MP $(CPPFLAGS) -DDEPENDS=1 $< >$*-tmp.dep && <$*-tmp.dep sed 's=\($*\)\.<b>dep</b>[ :]*=\1.o $@ : =' >$@ && rm $*-tmp.dep
</pre>

This appears to fix the e-directory's issues.  If no one sees any unintended consequences, I'll go ahead and do a pull request.

Also, the file Macaulay2/d/engine.o in the build directory does not get updated when it should, but this hasn't yet caused a problem for me (yet).  The dependencies for this file look very different and I don't know what should be changed to fix this.

**Comment by DanGrayson** [comment]
Thanks.  Do you know what the problem was, and why this fixes it?

**Comment by MichaelABurr** [comment]
> Thanks. Do you know what the problem was, and why this fixes it?

This is summarizing some ideas from meetings that I've had with @mikestillman and @antonleykin : 

1) For .dep files in subdirectories of the e-directory, the prefix of the path was missing.
    * For example, for ring.o in Macaulay2/e/interface, the dep file's had ``ring.o ring.dep`` instead of ``interface/ring.o interface/ring.dep``
    * The command ``-MT"$@"`` includes the directory structure.

2) Including the ``-MT"$@"`` had an unexpected consequence that the name being passed around in the makefile command changed from a .o file to a .dep file (at least this is what seemed to appear in the temp file, but I'm a little suspicious of this description).
     * For example, in the original, the dependency started with ``ring.o``, but after the ``-MT"$@"``, the file started with ``ring.dep``
     * I changed the ``.o`` to ``.dep`` in the regular expression in the ``sed`` command in the makefile.  (I'm concerned that this change may have unintended consequences since I'm not sure how complicated these dependency files can get.)

**Comment by DanGrayson** [comment]
Thanks for discovering the `-MT` option.  It appears that it can also be used multiple times to add dependencies, so we can use that feature to eliminate the `sed` step, as in the diff below.  I think we should also be more careful to have all the command line options that same as when compiling, and so I've added more flags.  Can you try that out, too?  It seems to work here.

```diff
diff --git a/M2/Macaulay2/e/Makefile.common.in b/M2/Macaulay2/e/Makefile.common.in
index a4a940985..d2b8b7300 100644
--- a/M2/Macaulay2/e/Makefile.common.in
+++ b/M2/Macaulay2/e/Makefile.common.in
@@ -66,12 +66,12 @@ COMPILE.cpp = $(COMPILE.cc)
 #	$(WHY)
 	$(SHOW)"MKDEP $*.cpp"
 	$(HIDE)$(MKDIR_P) `dirname "$@"`
-	$(HIDE)@CXX@ -MM -MP $(CPPFLAGS) $(CXXFLAGS) -DDEPENDS=1 $< >$*-tmp.dep && <$*-tmp.dep sed 's=\($*\)\.o[ :]*=\1.o $@ : =' >$@ && rm $*-tmp.dep
+	$(HIDE)@CXX@ -MT $@ -MT $*.o -MM -MP $(CXXFLAGS) $(CPPFLAGS) $(TARGET_ARCH) $(MORE_OPTIONS) $< >$*.dep
 
 %.dep: %.c
 #	$(WHY)
 	$(SHOW)"MKDEP $*.c"
-	$(HIDE)@CC@  -MM -MP $(CPPFLAGS) -DDEPENDS=1 $< >$*-tmp.dep && <$*-tmp.dep sed 's=\($*\)\.o[ :]*=\1.o $@ : =' >$@ && rm $*-tmp.dep
+	$(HIDE)@CC@  -MT $@ -MT $*.o -MM -MP $(CFLAGS)   $(CPPFLAGS) $(TARGET_ARCH) $(MORE_OPTIONS) $< >$*.dep
 
 # Local Variables:
 # compile-command: "make -C $M2BUILDDIR/Macaulay2/e DEPENDS=no initialize && make -C $M2BUILDDIR/Macaulay2/e Makefile.common "
```

**Comment by MichaelABurr** [comment]
> Can you try that out, too? It seems to work here.

It appears that your approach is cleaner and produces essentially the same result (I checked both Macaulay2/e/ring.dep and Macaulay2/e/interface/ring.dep)

**Comment by DanGrayson** [comment]
Thanks for testing it.  A pull request would be welcome.

**Comment by MichaelABurr** [comment]
Closed by pull request #1994
